### PR TITLE
Unit hydrograph + hourly timestep + more realistic PDFs

### PR DESCRIPTION
**Pull request by quinnylee**
### Incorporated unit hydrograph functionality
- Used [the pyflo package](https://github.com/benjiyamin/pyflo) to transform `q_spigot + q_overflow` using the unit hydrograph transformation, producing an output called `q_total`. 
- Pyflo is installed with the command `$ pip install pyflo`. A distribution .csv file is also required to determine the shape of the unit hydrograph. The default is found in `distributions/scs484.csv`, and it is defined as follows: `uh484 = system.array_from_csv('distributions/scs484.csv')`
- The default distribution (peak rate factor 484) was kept. It would be possible to change this, as well as other parameters like time of concentration and a parameter controlling direct runoff, but I didn't feel like I understood these well enough to change them.

### Changed timestep to hour
- Allows for longer predictions and easier deployment
- Transformed `g` to be in $\frac{m}{h^{2}}$

### Normalized q values by dividing by basin area
- Outputs are now yields, not discharges. This keeps them in the same unit (m) as the input, precipitation.
- Pyflo outputs everything in $\frac{ft^{3}}{s}$. These had to be transformed into $\frac{m}{h}$.

### Incorporated PDFs of geological permeability, ET, and precip for more realistic value generation
- Permeability (log 10) generated using a Normal(-13.8857, 1.1835) distribution
download-1
- ET generated using a Weibull_min(2.2447, 9.9807e-5, 0.0016) distribution
download
- Light precip generated using a Gumbel_r(0.0008108, 0.0009759) distribution
download-4
- Heavy precip generated using a GenPareto(0.2341, 0.0101, 0.009250) distribution
download-5
- PDFs fitted using a selection of PDFs identified during literature review, fitting with both the maximum likelihood and method of moments techniques, and goodness of fit evaluated using Kolmogorov-Smirnov test. All calculations done using scipy.stats package

### Updated infiltration calculation 
- My original interpretation of Darcy's law drained the bucket in a suspiciously fast way
- New calculation: infiltration = $\frac{10^{K} * (hwaterlevel + soildepth)}{soildepth}$

### Updated A_spigot and H_spigot calculations to be a function of H_bucket
- Bucket parameters now include `rA_spigot` and `rH_spigot`.
- `A_spigot` calculated as area of a circle with diameter `rA_spigot * H_bucket`
- `H_spigot` calculated as `rH_spigot * H_bucket`

### Updated model validation and results visualization function
- Included `q_total` as an output for visualization and validation 
- `q_spigot` and `q_overflow` still included in bucket output plot and LSTM outputs
- Static attributes removed from bucket visualization for clarity. Information printed above plots instead
- `h_bucket` plotted in separate subplot for clarity/scale issues
- Calculates NSE for `q_total` now

### Updated water balance calculation
- `infiltration` and `q_total` added to `mass_out`
- `q_spigot` and `q_overflow` removed from `mass_out`

### Example bucket outputs: 
image

### Example LSTM result visualizations:
image

### Tasks for the future:
- Currently the buckets are very wet. We could adjust the PDFs or the Markov chain to create a somewhat wet bucket.
- Initial values of `h_bucket` might significantly affect bucket results. `h_bucket[0]` might need to be incorporated as an input in the LSTM

### CODE REVIEW

**Comment by jmframe** [comment]
Can you include instructions on how to install pyflo? It isn't a simple pip install? Do we have to clone this and install from the source: https://github.com/benjiyamin/pyflo?

**Comment by quinnylee** [comment]
Yep, it's just a pip install, but it also requires a distribution .csv file. An example of one was added in [this commit](https://github.com/NWC-CUAHSI-Summer-Institute/deep_bucket_lab/pull/17/commits/e40175ea3eb296dac5fbc2e0cd3a823451c51b7b). I also added a line about referencing and defining the distribution in the above comment

**Comment by jmframe** [comment]
I'd like to get some better LSTM results before we merge this. There are a couple of things going on here preventing the LSTM from learning the response. I just made a couple of tweaks to the setup, like using a shorter time series, and a lower bucket, to get an overflow response quicker in the time series. 


image

**Comment by quinnylee** [comment]
Is it possible that including the initial value of h_bucket as an input would improve predictions? It seems to have a significant impact on the data generated. Although, that might be data contamination. What do you think?

**Comment by jmframe** [comment]
It probably would, but in a "real world" scenario, we wouldn't know the initial state values. We usually warm up models to get the state values in a reasonable configuration.

**Comment by quinnylee** [comment]
Okay, the first couple things coming to mind for improving the LSTM (on top of the two changes you made) would be increasing number of buckets produced, number of epochs, number of layers/hidden states... am I on the right track? Or should I be tweaking something in the bucket model or even the LSTM itself?

**Comment by jmframe** [comment]
You are on the right track, yes. The more complicated your system, the more complex your model needs to be, true. 

But what I was thinking is to try to get more of the total hydrologic response, with increasing and decreasing spigot flow and some overflows earlier in the simulation, so we can reduce the total training time steps down to like a few hundred, so we can add more buckets.

**Comment by jmframe** [comment]
Thanks @quinnylee! I am merging this PR, but then I am going to make a few changes immediately after. 

Then, we need to make the update to the main code in /src to match :)

File: deep_bucket_lab.ipynb
Comment (jmframe) [inline deep_bucket_lab.ipynb:1]: Will make some changes to the distribution of rainfall, and some of the model inputs, and a bit on the data loader to make it easier to train.

**Review by jmframe — APPROVED** [review]
Some further changes that need to be made, but will do following this PR. Thanks @quinnylee

**Pull request merged**